### PR TITLE
Migrate to fan-template platform

### PR DESCRIPTION
**Pull request by ireun**
https://esphome.io/components/fan/template
This allows to use https://github.com/denysdovhan/purifier-card

image

### CODE REVIEW

File: mipurifier.h [hunk 1]
```diff
@@ -109,19 +109,23 @@ class MiPurifier : public Component, public UARTDevice, public CustomAPIDevice {
...
+      } else {
+        XiaomiPurifierFan->turn_off();
+      }
     } else if (strcmp(id, "25") == 0) {
       // mode (auto, night, manual, preset)
       is_preset = false;
       switch (atoi(val)) {
         case 0:
-          mode_select->publish_state("auto");
+          XiaomiPurifierFan->preset_mode = "auto";
```
Comment (ireun) [inline mipurifier.h:122]: Note: changing mode using physical button on under the LCD does not work correctly.

When doing something like 
```cpp
auto call = XiaomiPurifierFan->make_call();
call.set_preset_mode("auto");
call.perform();
```
it is updated correctly, but for some reason this does not allow for turning the purifier off, it instantly turns on again.

@jaromeyer maybe you'll figure it out, for now I'm leaving this as a draft for others :)

**Comment by ireun** [comment]
Few commits that have implemented that platform:

https://github.com/esphome/esphome/pull/6310
https://github.com/esphome/esphome-docs/pull/3655
https://github.com/esphome/esphome/pull/6112

File: mipurifier.h (hunk 1 above)
Comment (ireun) [reply mipurifier.h]: Resolved with 

```cpp
    on_preset_set:
      - lambda: |-
          auto c = static_cast<MiPurifier *>(mipurifier);
          if (!id(XiaomiPurifierFan).state) {
            c->set_mode(x);
            c->turn_off();
          } else {
            c->set_mode(x);
          }
```

**Comment by jaromeyer** [comment]
sorry ive been very busy. i'll try to have a look at it in the following days

**Comment by Xploder** [comment]
For me, it doesn't sync the fan percentage to the actual value at all but having this setup as a fan entity is the right way in general.

**Comment by leonardpitzu** [comment]
It works but the fan speed slider has no effect. Does it work for you?

**Comment by ireun** [comment]
Have you set the mode to manual?

**Comment by leonardpitzu** [comment]
I guess so. Clicked the "fan and moved the slider up and down - nothing.

**Comment by ireun** [comment]
image

**Comment by leonardpitzu** [comment]
still nothing...

**Comment by ireun** [comment]
Will check if the code here is the same as mine, or weather I have made any changes. Though it works for me..

**Comment by leonardpitzu** [comment]
only looked at it from hi up in the sky but after
`XiaomiPurifierFan->speed = atof(val)+1;`
isn't there supposed to be a
`call.perform();`
?

**Comment by ireun** [comment]
I've found one some differences between what I use in HA and what was in this draft, could You try that?

**Comment by leonardpitzu** [comment]
it's on now but one suggestion - whenever the slider is moved just change the preset automatically to manual
in this case you do not need to firstly change the preset and then move the slider

**Comment by ireun** [comment]
I've tried to do that, but I wasn't able to achieve that, since than I have not worked on it, maybe @jaromeyer could figure it out ;)

**Comment by Xploder** [comment]
Yeah that'd be great. Right now if you're in let's say the Medium preset and change the percentage slider, the unit itself stays in that preset and doesn't actually change the speed and on Home-Assistant it goes to an unknown/bogus state for the mode selection

Before adjusting %:
grafik


After adjusting % from the medium preset:
grafik

**Comment by Xploder** [comment]
`    on_speed_set:
      - lambda: |-
          auto c = static_cast<MiPurifier *>(mipurifier);
          c->set_mode("manual");
      - delay: 1s
      - lambda: |-
          auto c = static_cast<MiPurifier *>(mipurifier);
          c->set_manualspeed((int)x-1);`

I've changed the on_speed_set function to set manual mode every time I change the speed

**Comment by ireun** [comment]
Thanks @Xploder! I've swapped these lambada, so first the speed is set and than the mode, to prevent going it to a previous speed for that split second :)

**Comment by Xploder** [comment]
Thanks! Works fine within the usual limitations of it not being able to read current mode later etc.
But now we're at least able to easily set percentage from Alexa from any mode!

**Comment by leonardpitzu** [comment]
works great. like this, enough for getting things done

**Comment by jaromeyer** [comment]
Finally got to test your PR and it works great! Thanks @ireun

**Pull request merged**